mesonbuild/dependencies/base.py

```diff
@@ -355,6 +363,28 @@ def generate_link_whole_dependency(self) -> Dependency:
         new_dep.libraries = []
         return new_dep
 
+    def get_as_static(self, recursive: bool) -> Dependency:
+        new_dep = copy.deepcopy(self)
+        new_dep.prefer_static = True
+        if recursive:
+            new_dep.ext_deps = [dep.get_as_static(True) if isinstance(dep, InternalDependency) else dep for dep in new_dep.ext_deps]
+        return new_dep
+
+    def get_as_shared(self, recursive: bool) -> Dependency:
+        new_dep = copy.deepcopy(self)
+        new_dep.prefer_static = False
+        if recursive:
+            new_dep.ext_deps = [dep.get_as_shared(True) if isinstance(dep, InternalDependency) else dep for dep in new_dep.ext_deps]
+        return new_dep
+
+    def get_libraries(self) -> T.Generator[LibTypes, None, None]:
+        from ..build import BothLibraries
+        for lib in self.libraries:
+            if isinstance(lib, BothLibraries):
+                yield lib.static if self.prefer_static else lib.shared
+            else:
+                yield lib
```

Comment:
I'm a bit scared that we could have multiple places where we do `target.libraries` instead of using the getter. I haven't checked the code in details, are you sure it can't happen?

Maybe safer to modify `target.libraries` directly in `get_as_static()` method?

Reply:
Yes, I will try that. I think it will simplify a lot of things...